### DISCUSSION TITLE
Multiple keyboard keys / joystick repeat event?

### DISCUSSION THREAD

**Discussion by FlorentLM** [zauberzeug/nicegui#333] [Q&A] [answered]
Hi there,

I am currently experimenting nicegui by building a simple 2D game with svg sprites. I had a lot of fun so far, nicegui is extremely satisfying to use!

I tried both controlling my player with the keyboard and with a vitrual joystick - that works but I have a small issue in both cases:
- The keyboard control is only detecting one key at a time, so it is impossible to move diagonally, and it seems that the polling interval is overall quite slow?
- The joystick works much much more smoothly, but it seems to only respond when moved, not when held in a direction.

I couldn't find any help about this in the documentation, but I must be missing something?

Many thanks!

**Comment by falkoschindler** [comment, marked as answer]
Both elements fire events only when something changes. Key events are closely coupled to the corresponding JavaScript events: https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent. They don't tell you what keys are currently pressed, but which key is changing its state, either down or up. If you need to know all currently pressed keys, you need to keep track of them yourself, unfortunately. The joystick behaves similarly: It only fires on change. Otherwise it would need to constantly send events at enormous (actually infinite) frequency.

When implementing a game, you usually have some game loop that checks inputs and updates the output in regular intervals. You could, e.g., use a `ui.timer` for that. Let `ui.keyboard` and `ui.joystick` update some global state with a list of currently pressed keys and the current joystick coordinates, and let the timer consume this information to advance the game logic.

**Reply by FlorentLM** [reply]
Alright, thanks for the clarification!
I did what you suggested and it works flawlessly :)

Thanks!
